text/2020-11-09-chaos-mesh-workflow.md

```diff
@@ -0,0 +1,290 @@
+turn into `DeadlineExceeded` status.
+
+#### `Serial`
+
+For a `Serial` task, the engine should run these tasks one by one. It enters the
+next step iff the former one succeeded. After all these steps finished, this
+task turns into the finished phase.
+
+If one of the tasks failed or exceeded the deadline, the following task will not
+run and the current task will turn into the corresponding status directly.
+
+If the `deadline` field is set, all the running child task will be killed and
+turn into `DeadlineExceeded` status.
+
+#### `Suspend`
+
+For a `Suspend` task, the engine should do nothing. After a period of time
+(specified by `duration` field), this task turns into the finished phase.
+
+#### `Task`
+
+`Task` is a special kind of template to integrate users' processes into the
+workflow. In this step, the users can setup a container (with Kubernetes's pod)
+to run their own process. This step finishes once the pod turns into
+`PodSucceeded` phase. The description of the task could be more complicated with
+more fields if needed in the future. Every branch under the `branch` field will
+run parallelly iff the `when` expression returns true. There will be a lot of
+variables provided by the engine to use in the `when` expression, such as
+`stdout`, `stderr`...
+
+Combining `Parallel` and `Serial` can provide a powerful expression in task
+combination. However, as it can only represent series-parallel graph, it's still
+a subset of `dag`.
+
+#### Status
+
+The status of a `Workflow` and the transition between different status could be
+the heart of a workflow engine. Here is an example:
+
+```yaml
+status:
+  nodes:
+  - name: entry
+    displayName: serial-chaos
+    type: Serial
+    children:
+    - serial-chaos[0]
+    - serial-chaos[1]
+    startTime: Thu, 05 Nov 2020 14:23:25 +0800
+    deadline: 50m
+    phase: Running
+    spec:
+    - ref: networkchaos
+    - ref: iochaos
+    - ref: parallel-chaos
+    - type: Suspend
+      duration: "30m"
+    - ref: check
+  - name: entry[0]
+    displayName: networkchaos
+    type: NetworkChaos
+    startTime: Thu, 05 Nov 2020 14:23:25 +0800
+    duration: 10m
+    phase: Succeeded
+    spec:
+      selector:
+        labelSelectors:
+          "component": "tikv"
+      delay:
+        latency: "90ms"
+        correlation: "25"
+        jitter: "90ms"
+  - name: entry[1]
+    displayName: parallel-chaos
+    type: Parallel
+    startTime: Thu, 05 Nov 2020 14:33:25 +0800
+    deadline: 10m
+    phase: Running
+    spec:
+    - ref: networkchaos
+    - ref: iochaos
+  - name: entry[1][0]
+    displayName: networkchaos
+    type: NetworkChaos
+    startTime: Thu, 05 Nov 2020 14:33:25 +0800
+    duration: 10m
+    phase: Running
+    spec:
+      selector:
+        labelSelectors:
+          "component": "tikv"
+      delay:
+        latency: "90ms"
+        correlation: "25"
+        jitter: "90ms"
+  - name: entry[1][1]
```

Comment:
Does the `[1][1]` means the execution order? Would the real world instance also keep this naming?

Reply:
Because `entry` is a `Serial` template, `entry[1]` means its first task, which is an instance of `parallel-chaos`. And as `parallel-chaos` is a `Parallel`, `entry[1][1]` means its first task, which is an instance of `iochaos` template.

The name doesn't suggest the execution order. They are only an index to be searched by their parents so that their parents can check the status of children easily.

Reply:
Okay. Another question is, do we need some more workflow-level fields? There are ONLY nodes inside the workflow status. Looking at [the argo workflow status fields](https://argoproj.github.io/argo/fields/#workflowstatus), it did have some fields more, e.g. :

```
conditions:  Conditions is a list of conditions the Workflow may have
finishedAt:  Time at which this workflow completed
phase:  Phase a simple, high-level summary of where the workflow is in its lifecycle.
startedAt: Time at which this workflow started
...
```

We may also need these fields to show the high-level summary of the workflow state, WDYT?

Reply:
Yes! I agree with you 🍻